Theme Sheet: Add Tracks events to Support and CSS buttons

jeremeylduvall opened the pull request:
This is a fix for #10262. It adds Tracks events to both the Support and CSS
links within a theme sheet. The event is recorded in Tracks along with the theme
name.

## To test
1. Load this branch and visit http://calypso.localhost:3000/
2. Click "My Sites" -> Themes
3. Select a theme from the list
4. From the theme sheet, click the "Support" tab
5. Click either the "Need extra help?" Contact Us button or the "Need CSS help?" Visit Forum button.

You can use `localStorage.setItem('debug', 'calypso:analytics');` to log out the Tracks events to your console. You should see either `calypso_theme_help_button_click` or `calypso_theme_css_button_click` called with the property `theme_name` passed in.

jeremeylduvall commented:
Instead of separating out the Tracks calls, I combined them into one call passing in `Context` so we could tell which button was clicked:

```
const mapDispatchToProps = ( dispatch, ownProps ) => ( {
	recordThemeSheetButtonClick: ( context ) => dispatch( recordTracksEvent( 'calypso_theme_sheet_button_click', {
		theme_name: ownProps.id,
		button_context: context
	} ) )
} );
```

The only issue I would like to clear up is the ESLint warning `JSX props should not use .bind()`. I understand why, but I'm not quite sure how to get around this and still pass in the context of the button clicked to `recordThemeSheetButtonClick()`. 

I tried separating it out like this:

```
handleClick( context ) {
	this.props.recordThemeSheetButtonClick( context );
}

//...

renderCssSupportCard() {
	return (
		//...
			<Button
				href="//en.forums.wordpress.com/forum/css-customization"
				onClick={ this.handleClick.bind( null, 'css' ) }>
				Visit forum
			</Button>
		//...		
	);
},

//...

const mapDispatchToProps = ( dispatch, ownProps ) => ( {
	recordThemeSheetButtonClick: ( context ) => dispatch( recordTracksEvent( 'calypso_theme_sheet_button_click', {
		theme_name: ownProps.id,
		button_context: context
	} ) )
} );
```

But I'm still running into the same warning. Any advice much appreciated!

obenland commented:
The easiest way short of converting it to an ES6 class is probably to define a method for each use case.

```js
const ThemeSheet = React.createClass( {
    // ...

	trackButtonClick( context ) {
		this.props.recordTracksEvent( 'calypso_theme_sheet_button_click', {
			theme_name: this.props.id,
			button_context: context,
		} )
	},

	trackHelp() {
		this.trackButtonClick( 'help' );
	},

	trackCss() {
		this.trackButtonClick( 'css' );
	},

	// ...
} );

connect(
	( state, { id } ) => {
		// ...
	},
	{
		recordTracksEvent,
	},
)( ThemeSheetWithOptions )
```

In a class we could define that in the constructor and make it a little prettier:
```js
class ThemeSheet extends Component {
	constructor() {
		super( ...arguments );

		this.trackHelp = this.trackButtonClick.bind( this, 'help' );
		this.trackCss = this.trackButtonClick.bind( this, 'css' );

		this.state = { showPreview: false };
	}

	trackButtonClick( context ) {
		this.props.recordTracksEvent( 'calypso_theme_sheet_button_click', {
			theme_name: this.props.id,
			button_context: context,
		} );
	}

	// ...
} );

connect(
	( state, { id } ) => {
		// ...
	},
	{
		recordTracksEvent,
	},
)( ThemeSheetWithOptions )
```

I wonder if @gwwar can think of more/better ways to do that?

gwwar commented:
@obenland's suggestion looks good to me. I'd recommend holding off on porting the class to es2015 into another janitorial PR that has no functional changes since this component looks larger.

jeremeylduvall commented:
Thanks @obenland! I elected to just define a separate method for each item and then rely on `this.props.recordThemeSheetButtonClick( 'help' );`. Still working as originally described.

obenland commented on client/my-sites/theme/main.jsx:
```diff
@@ -577,6 +595,13 @@ const ThemeSheetWithOptions = ( props ) => {
+		theme_name: ownProps.id,
+		button_context: context
+	} ) )
+} );
```
I guess technically there is nothing wrong to do it this way, just note that if we ever decided to use `recordTracksEvent` somewhere else in this component it couldn't be used right away but would need to get its own mapping here

obenland commented on client/my-sites/theme/main.jsx:
```diff
@@ -577,6 +595,13 @@ const ThemeSheetWithOptions = ( props ) => {
 	);
 };
 
+const mapDispatchToProps = ( dispatch, ownProps ) => ( {
```
Could this be inlined for consistency with `mapStateToProps`?

gwwar replied:
Plus 1 for consistency, though `mapStateToProps` is long enough here that we should probably pull that out into it's own function.

gwwar commented on client/my-sites/theme/main.jsx:
```diff
@@ -273,7 +282,12 @@ const ThemeSheet = React.createClass( {
+					primary={ isPrimary }
+					href={ '/help/contact/' }
+					onClick={ this.trackContactUsClick }>
+					Contact us
```
Missing translate here. We can probably also wrap the component in Localize though let us know if there's a reason not to.

gwwar commented on client/my-sites/theme/main.jsx:
```diff
@@ -307,7 +321,11 @@ const ThemeSheet = React.createClass( {
+				<Button
+					href="//en.forums.wordpress.com/forum/css-customization"
+					onClick={ this.trackCssClick }>
+					Visit forum
```
Missing translate here.

jeremeylduvall commented:
Thanks @obenland and @gwwar. I appreciate the help!

I like the idea of moving `mapStateToProps` into it's own function versus inlining `mapDispatchToProps`. When I pulled out `mapStateToProps` though, I received this error:

> Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the ThemesSelectionWithPage component.

I'm unclear exactly why this is happening. [Per the comment](https://github.com/Automattic/wp-calypso/blob/7db145c95136465de8d1fe362f1f310526fddd05/client/my-sites/theme/main.jsx#L606) in `main.jsx`, it's clear there's a relationship with `ThemesSelectionWithPage`. To workaround this, I had to create an `_isMounted` reference within `ThemesSelectionWithPage` ([from React docs](https://facebook.github.io/react/blog/2015/12/16/ismounted-antipattern.html)). I'm not sure if this is the best way forward.

Other improvements:
- This now uses `localize` instead of `this.translate`
- I swapped out how we were handling the Tracks call [per Obenland's comment](https://github.com/Automattic/wp-calypso/pull/10430#pullrequestreview-17279018)

a8ci18n commented on client/my-sites/theme/main.jsx:
```diff
@@ -300,14 +329,19 @@ const ThemeSheet = React.createClass( {
+				<Button
+					href="//en.forums.wordpress.com/forum/css-customization"
+					onClick={ this.trackCssClick }>
+					{ translate( 'Visit forum' ) }
```
Alternate string suggestion: 
* `translate( 'Visit WordPress Forum' )`&mdash; translations: **8**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/10262-tracks-theme-sheeets)

gwwar commented:
Maybe @ockham could help explain the context further. If you are seeing different behavior pulling the function out, it's safer to keep `mapStateToProps` inlined for now.

ockham commented:
Uh, that's really weird. I'll have a closer look later today.

ockham commented:
Sorry, wasn't able to get to this on Friday. Looking now.

ockham commented:
> @obenland's suggestion looks good to me. I'd recommend holding off on porting the class to es2015 into another janitorial PR that has no functional changes since this component looks larger.

That's #10748 BTW 😄

ockham commented:
> > Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the ThemesSelectionWithPage component.

> I'm unclear exactly why this is happening. [Per the comment](https://github.com/Automattic/wp-calypso/blob/7db145c95136465de8d1fe362f1f310526fddd05/client/my-sites/theme/main.jsx#L606) in `main.jsx`, it's clear there's a relationship with `ThemesSelectionWithPage`. To workaround this, I had to create an `_isMounted` reference within `ThemesSelectionWithPage` ([from React docs](https://facebook.github.io/react/blog/2015/12/16/ismounted-antipattern.html)). I'm not sure if this is the best way forward.

Yeah, I'd rather get to the bottom of this, since that `isMounted` check is more of a band-aid fix that'd conceal an underlying issue...

ockham commented:
@jeremeylduvall Are you sure the `setState` error is caused by the separate `mapDispatchToProps`? I seem to be able to repro it on `master` (start at `/design`, click on any theme screenshot), which would mean it's pre-existing rather than a regression caused by this PR.

ockham commented:
(If you can, let's drop the `isMounted` stuff and deal with the `setState` warning separately.)

budzanowski commented:
@jeremeylduvall Is it possible to limit this PR to just tracking?  #10748 PR addressees other elements you are implementing here like translate and ES6+. If it is not possible than OK I take the merge on me but I would like to avoid it. 10748 will need some more time before it is finished but this is my prio. now so work goes rather fast.

jeremeylduvall commented:
Thanks @budzanowski — I reverted this PR back to just the tracking changes. The only other thing I introduced was a `i18n.translate` call on the various buttons that weren't translated until we introduce `localize`. Let me know if that's alright.

Otherwise, this should still serve the originally purpose from the opening note. I verified that the actions are still recorded as expected.

@ockham 

> Are you sure the setState error is caused by the separate mapDispatchToProps?

I swear I tried this on master and didn't experience the error so I thought they were related. Now though, I see it on master as well. I've reverted those changes anyway.

a8ci18n commented on client/my-sites/theme/main.jsx:
```diff
@@ -296,7 +317,11 @@ const ThemeSheet = React.createClass( {
+				<Button
+					primary={ isPrimary }
+					href={ this.props.forumUrl }>
+					{ i18n.translate( 'Visit forum' ) }
```
Alternate string suggestion: 
* `translate( 'Visit WordPress Forum' )`&mdash; translations: **8**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/10262-tracks-theme-sheeets)

ockham commented:
> Thanks @budzanowski — I reverted this PR back to just the tracking changes. The only other thing I introduced was a `i18n.translate` call on the various buttons that weren't translated until we introduce `localize`. Let me know if that's alright.

Sounds good to me, since this was an oversight of ours that we should fix -- so it's okay if we have to rebase here.

> Otherwise, this should still serve the originally purpose from the opening note. I verified that the actions are still recorded as expected.

@ockham 

> > Are you sure the setState error is caused by the separate mapDispatchToProps?

> I swear I tried this on master and didn't experience the error so I thought they were related. Now though, I see it on master as well. I've reverted those changes anyway.

Cool, thanks. Giving this a final test now.

ockham commented:
When clicking the "Need CSS help?" Visit Forum button, I don't see any props passed:

![image](https://cloud.githubusercontent.com/assets/96308/22295539/935b66d6-e317-11e6-8a70-dfdb638be11f.png)

(Note the arrow in the screenshot points down, indicating that the object is expanded -- yet empty.)

ockham commented:
Note that if a user is on a plan, there's a third button shown (second one in this image):

![image](https://cloud.githubusercontent.com/assets/96308/22295694/18597684-e318-11e6-8c39-885792e75f17.png)

That's [this button](https://github.com/Automattic/wp-calypso/blob/f29bc0d24789846770cd65bc61246cb4ae903fe8/client/my-sites/theme/main.jsx#L299), which I guess should also get tracking attached to it.

a8ci18n commented on client/my-sites/theme/main.jsx:
```diff
@@ -296,7 +321,12 @@ const ThemeSheet = React.createClass( {
+					primary={ isPrimary }
+					href={ this.props.forumUrl }
+					onClick={ this.trackThemeForumClick }>
+					{ i18n.translate( 'Visit forum' ) }
```
Alternate string suggestion: 
* `translate( 'Visit WordPress Forum' )`&mdash; translations: **8**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/10262-tracks-theme-sheeets)

jeremeylduvall commented:
Added the tracking for the theme forum and changed up the context naming convention just a tad to differentiate the CSS forum from theme forum. Props are still being recorded on my side:

<img width="1009" alt="screen shot 2017-01-25 at 8 25 28 am" src="https://cloud.githubusercontent.com/assets/7240478/22296604/ee081b32-e2d7-11e6-8297-cf12d69da3f9.png">

Trying to think why this would happen.

ockham commented:
Hmm, interesting. But it's good that they _are_ recorded; I can confirm that I see mine in Tracks, too. So that's not a blocker then.
